subsys/bluetooth/mesh/scheduler_srv.c

```diff
@@ -0,0 +1,630 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <stdio.h>
+#include <bluetooth/mesh/models.h>
+#include <bluetooth/mesh/scheduler_srv.h>
+#include <bluetooth/mesh/scene_srv.h>
+#include <bluetooth/mesh/gen_onoff_srv.h>
+#include <sys/byteorder.h>
+#include <random/rand32.h>
+#include "model_utils.h"
+#include "mesh/net.h"
+#include "mesh/access.h"
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_MESH_DEBUG_MODEL)
+#define LOG_MODULE_NAME bt_mesh_scheduler_srv
+#include "common/log.h"
+
+#define ANY_YEAR       0x64
+#define ANY_DAY        0x00
+#define MAX_DAY        0x1F
+#define JANUARY           0
+#define DECEMBER         11
+#define ANY_HOUR       0x18
+#define RANDOM_HOUR    0x19
+#define ANY_UNIT       0x3C
+#define EVERY_15_UNITS 0x3D
+#define EVERY_20_UNITS 0x3E
+#define RANDOM_UNIT    0x3F
+
+static bool set_year(struct tm *sched_time,
+		     struct tm *current_local,
+		     struct bt_mesh_schedule_register *item);
+static bool set_month(struct tm *sched_time,
+		      struct tm *current_local,
+		      struct bt_mesh_schedule_register *item);
+static bool set_day(struct tm *sched_time,
+		    struct tm *current_local,
+		    struct bt_mesh_schedule_register *item,
+		    struct bt_mesh_time_srv *srv);
+static bool set_hour(struct tm *sched_time,
+		     struct tm *current_local,
+		     struct bt_mesh_schedule_register *item,
+		     struct bt_mesh_time_srv *srv);
+static bool set_minute(struct tm *sched_time,
+		       struct tm *current_local,
+		       struct bt_mesh_schedule_register *item,
+		       struct bt_mesh_time_srv *srv);
+static bool set_second(struct tm *sched_time,
+		       struct tm *current_local,
+		       struct bt_mesh_schedule_register *item,
+		       struct bt_mesh_time_srv *srv);
+
+static inline bool is_leap_year(uint32_t year)
+{
+	return ((year % 4) == 0) &&
+	       (((year % 100) != 0) || ((year % 400) == 0));
+}
+
+static bool revise_year(struct tm *sched_time,
+			struct tm *current_local,
+			struct bt_mesh_schedule_register *item,
+			int number_years)
+{
+	struct tm revised_time = *current_local;
+
+	revised_time.tm_year += number_years;
+	return set_year(sched_time, &revised_time, item);
+}
+
+static bool revise_month(struct tm *sched_time,
+			 struct tm *current_local,
+			 struct bt_mesh_schedule_register *item,
+			 int number_days)
+{
+	int days[12] = {31, is_leap_year(current_local->tm_year) ? 29 : 28,
+			31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
+	struct tm revised_time = *current_local;
+
+	revised_time.tm_mday += number_days;
+
+	if (revised_time.tm_mday > days[revised_time.tm_mon]) {
+		revised_time.tm_mday -= days[revised_time.tm_mon];
+		revised_time.tm_mon = revised_time.tm_mon == DECEMBER ?
+				JANUARY : revised_time.tm_mon + 1;
```

Comment:
In this case, revised_time needs to update its year as well. Does it?

Reply:
you're right, it does.